### PR TITLE
define and use unsafeSetImmediate

### PR DESCRIPTION
**Pull request by safareli**

### CODE REVIEW

File: src/Routing/PushState.purs [hunk 1]
```diff
@@ -212,3 +200,38 @@ matchesWith parser cb = foldPaths go (go Nothing)
...
       <<< parser
+
+-- | Similar to `setImmediate`, it's impelemnted using microtask queue via MutationObserver
+-- | to schedule callbacks. this way it's more imediate then `setTimout` would have been.
+-- | We use a fresh counter so that the text change mutation always fires.
+-- |
+-- | NOTE: it's not that unsafe, but we use `unsafeRunRef` and `runPure` so it's still unsafe.
+unsafeSetImmediate
+  :: forall r
+  . Eff (ref :: REF, dom :: DOM |r) Unit -> Eff (ref :: REF, dom :: DOM |r) Unit
```
Comment (natefaubion) [inline src/Routing/PushState.purs]: Why use unsafe and runPure? https://github.com/natefaubion/purescript-spork/blob/master/src/Spork/Scheduler.purs#L20

We are already in Eff, creating a PushStateInterface. There's no reason to have a global `unsafeSetImmediate` created with unsafe effects. the microtask queue is already global.

Comment (safareli) [reply src/Routing/PushState.purs]: This way one node is used for all calls to `unsafeSetImmediate`.  but for this case it's not important

File: src/Routing/PushState.purs [hunk 2]
```diff
@@ -95,35 +95,18 @@ makeInterface = do
...
   let
     stateFn op state path = do
       DOM.window
         >>= DOM.history
         >>= op state (DOM.DocumentTitle "") (DOM.URL path)
+      -- The hashchange interface is asynchronous, since hashchange events are
+      -- fired on the next tick of the event loop. We want the push-state
+      -- interface to behave as similarly as possible, so we use something like
+      -- `setImmediate` and use `Ref Boolean` to make sure maximum one `notify` is
+      -- scheduled per event loop.
```
Comment (natefaubion) [inline src/Routing/PushState.purs]: I think this comment needs to be updated. We don't use a Ref Boolean here.

File: src/Routing/PushState.purs [hunk 3]
```diff
@@ -212,3 +195,27 @@ matchesWith parser cb = foldPaths go (go Nothing)
     maybe (pure a) (\b -> Just b <$ cb a b)
       <<< indexl 0
       <<< parser
+
+-- | Similar to `setImmediate`, it's impelemnted using microtask queue via MutationObserver
```
Comment (natefaubion) [inline src/Routing/PushState.purs]: Typo `implelemnted`

File: src/Routing/PushState.purs [hunk 4]
```diff
@@ -212,3 +195,27 @@ matchesWith parser cb = foldPaths go (go Nothing)
...
+-- | to schedule callbacks. this way it's more imediate then `setTimout` would have been.
```
Comment (natefaubion) [inline src/Routing/PushState.purs]: Capitalization `this`, and typo `imediate`.

**Pull request merged**